dali-mqtt-daemon.py

```diff
@@ -172,22 +188,44 @@ def main_loop(driver, max_lamps, mqtt_server, mqtt_port, ha_prefix):
               }
 
     try:
-        with open(args.config, 'r') as stream:
-            logger.debug("Loading configuration from <%s>", args.config)
-            config = yaml.load(stream)
-
-        MQTT_BASE_TOPIC = config["mqtt_base_topic"]
-
-        main_loop(config["dali_driver"], config["dali_lamps"], config["mqtt_server"], config["mqtt_port"], config["ha_discover_prefix"])
+        config = load_config_file(args.config)
+
+        driver_object = None
+        driver = args.dali_driver
+        logger.debug("Using <%s> driver", driver)
+        if driver == HASSEB:
+            from dali.driver.hasseb import SyncHassebDALIUSBDriver 
+            driver_object = SyncHassebDALIUSBDriver()
+        elif driver == TRIDONIC:
+            from dali.driver.tridonic import SyncTridonicDALIUSBDriver
+            driver_object = SyncTridonicDALIUSBDriver()
+        elif driver == DALI_SERVER:
+            from dali.driver.daliserver import DaliServer
+            driver_object = DaliServer("localhost", 55825)    
+
+        mqqtc = create_mqtt_client(driver_object, config["dali_lamps"], config["mqtt_server"], config["mqtt_port"], config["mqtt_base_topic"], config["ha_discover_prefix"])
+        mqqtc.loop_start()
+
+        watchdog_observer = Observer()
+        watchdog_event_handler = ConfigFileSystemEventHandler(mqqtc, config, args.config, driver_object, loop)
+        watchdog_observer.schedule(watchdog_event_handler, args.config)
+        watchdog_observer.start()
     except FileNotFoundError as e:
-        logger.info("Configuration file %s created, please reload daemon", args.config)
-    except KeyError as e:
-        missing_key = e.args[0]
-        config[missing_key] = args.__dict__[missing_key]
-        logger.info("Configuration file updated, please reload daemon")
-    finally:
         try:
             with io.open(args.config, 'w', encoding="utf8") as outfile:
                 yaml.dump(config, outfile, default_flow_style=False, allow_unicode=True)
+                logger.info("Configuration file %s created, please reload daemon", args.config)
         except Exception as err:
-            logger.error(f"Could not save configuration: {err}")
+            logger.error(f"Could not save configuration: {err}")   
+    while True:
```

Comment:
This makes no sense... just use mqttc loop, no need for an extra loop!

Reply:
Function [loop_forever](https://github.com/eclipse/paho.mqtt.python#loop_forever) is blocking main thread, I can use `loop_forever` instead of this infinite loop but it's create a problem, when I want to reload script on-fly. `disconnect` can end loop created by `loop_forever`, but before I can start new `loop_forever` script reaches end (because job which is blocking main thread is ended by `disconnect`).

So, to reload script on-fly I create a loop by `loop_start` and end loop by `loop_stop`, what isn't blocking main thread and I use infinite loop to prevent reaching end of main function

Reply:
from loop_forever documentation:

"If you call disconnect() in a callback it will return."

if you place loop_forever inside a guard, you will be able to check if you

Another approach is to use https://github.com/eclipse/paho.mqtt.python#loop, but it will not handle reconnect autonomously

Reply:
I'm not sure that I correctly understand your suggestion, I'm restored `loop_forever` and moved a code to reload configuration

Reply:
> from loop_forever documentation:
> "If you call disconnect() in a callback it will return."
@dgomes  I can't find this information in the doc, can you attach link to doc?

Reply:
https://github.com/eclipse/paho.mqtt.python/blob/master/src/paho/mqtt/client.py#L1749